TESTS/mbed_drivers/timerevent/main.cpp

```diff
@@ -0,0 +1,239 @@
+
+    // Set relative timestamp of internal event to present_time + ts
+    void set_future_timestamp(timestamp_t ts) {
+        insert(::ticker_read(_ticker_data) + ts);
+    }
+
+    void set_past_timestamp(void) {
+        insert(::ticker_read(_ticker_data) - 1UL);
+    }
+
+    void set_timestamp(timestamp_t ts) {
+        insert(ts);
+    }
+};
+
+class TestTimerEventAbsolute: public TestTimerEvent {
+public:
+    static const int32_t SEM_SLOTS_AFTER_ZERO_TS_INSERTED = 1;
+    static const int32_t SEM_SLOTS_AFTER_PAST_TS_INSERTED = 1;
+    TestTimerEventAbsolute() :
+            TestTimerEvent() {
+    }
+
+    TestTimerEventAbsolute(const ticker_data_t *data) :
+            TestTimerEvent(data) {
+    }
+
+    // Set absolute timestamp of internal event to present_time + ts
+    void set_future_timestamp(us_timestamp_t ts) {
+        insert_absolute(::ticker_read_us(_ticker_data) + ts);
+    }
+
+    void set_past_timestamp(void) {
+        insert_absolute(::ticker_read_us(_ticker_data) - 1ULL);
+    }
+
+    void set_timestamp(us_timestamp_t ts) {
+        insert_absolute(ts);
+    }
+};
+
+/** Template for tests: insert, insert_absolute
+ *
+ * Test insert
+ * Given an instance of @a TimerEvent subclass
+ * When a tiestamp is set with @a insert()
+ *     and given time elapses
+ * Then an event handler is called
+ *
+ * Test insert_absolute
+ * Given an instance of @a TimerEvent subclass
+ * When a tiestamp is set with @a insert_absolute()
+ *     and given time elapses
+ * Then an event handler is called
+ */
+template<typename T>
+void test_insert(void) {
+    T tte;
+
+    tte.set_future_timestamp(TEST_DELAY_US);
+    int32_t sem_slots = tte.sem_wait(0);
+    TEST_ASSERT_EQUAL(0, sem_slots);
+
+    sem_slots = tte.sem_wait(TEST_DELAY_US / 1000 + 1);
+    TEST_ASSERT_EQUAL(1, sem_slots);
+
+    tte.remove();
+}
+
+/** Template for tests: remove
+ *
+ * Test remove after insert
+ * Given an instance of @a TimerEvent subclass
+ * When a tiestamp is set with @a insert()
+ *     and timestamp is removed before being reached
+ * Then the event handler is never called
+ *
+ * Test remove after insert_absolute
+ * Given an instance of @a TimerEvent subclass
+ * When a tiestamp is set with @a insert_absolute()
+ *     and timestamp is removed before being reached
+ * Then the event handler is never called
+ */
+template<typename T>
+void test_remove(void) {
+    T tte;
+
+    tte.set_future_timestamp(TEST_DELAY_US * 2);
+    int32_t sem_slots = tte.sem_wait(TEST_DELAY_US / 1000);
+    TEST_ASSERT_EQUAL(0, sem_slots);
+    tte.remove();
+
+    sem_slots = tte.sem_wait(TEST_DELAY_US * 2 / 1000 + 1);
+    TEST_ASSERT_EQUAL(0, sem_slots);
+}
+
+/** Template for tests: insert, insert_absolute 0 us timestamp
+ *
+ * Test insert zero
+ * Given an instance of @a TimerEvent subclass
+ * When a timestamp of 0 us is set with @a insert()
+ * Then an event handler is **not** called instantly
+ *     (the event is scheduled after the ticker's overflow)
```

Comment:
What if a timestamp of 0 us is set with insert() when the lower 32 bits of the present time is 0? Shall we fire instantly or wait for the next overflow?

```
static us_timestamp_t convert_timestamp(us_timestamp_t ref, timestamp_t timestamp)
{
    bool overflow = timestamp < ((timestamp_t) ref) ? true : false;

    us_timestamp_t result = (ref & ~((us_timestamp_t)UINT32_MAX)) | timestamp;
    if (overflow) { 
        result += (1ULL<<32);
    }

    return result;
}
```
According to the code snippet from mbed_ticker_api.c, it will fire instantly.

Reply:
> According to the code snippet from mbed_ticker_api.c, it will fire instantly.

There is only one case when `overflow` flag will not be set to `true` for `timestamp = 0` -- that is when `ref = 0`. That should never happen since us_ticker is initialized before the call to `TimerEvent::insert()`.

Reply:
```
bool overflow = timestamp < ((timestamp_t) ref) ? true : false;
```
If `timestamp==0`, `overflow` will be true whenever the lower 32 bits of `ref` is 0.

Reply:
@jiandeng Thanks for pointing this out, you're right. I've somehow overlooked the type cast.

This test case does not make much sense to me now. I'll need to update the code